### PR TITLE
GC safe attribute access

### PR DESCRIPTION
**Pull request by UnityAlex**

### CODE REVIEW

File: unity/unity-sources/Runtime/Mono/MonoFunctions.h [hunk 1]
```diff
@@ -312,25 +312,13 @@ DO_API(void, mono_verifier_set_mode, (MiniVerifierMode mode))
...
-DO_API(MonoCustomAttrInfo*, mono_custom_attrs_from_method, (MonoMethod * method))
-DO_API(MonoCustomAttrInfo*, mono_custom_attrs_from_property, (MonoClass * klass, MonoProperty * property))
-DO_API(MonoCustomAttrInfo*, mono_custom_attrs_from_class, (MonoClass * klass))
-DO_API(MonoCustomAttrInfo*, mono_custom_attrs_from_assembly, (MonoAssembly * assembly))
 DO_API(MonoArray*, mono_reflection_get_custom_attrs_by_type, (MonoObject * object, MonoClass * klass))
-DO_API(void, mono_custom_attrs_free, (MonoCustomAttrInfo * attr))
-DO_API(gboolean, mono_unity_class_has_attribute, (MonoClass* klass, MonoClass* attr_klass))
-DO_API(gboolean, mono_unity_assembly_has_attribute, (MonoAssembly* assembly, MonoClass* attr_klass))
-DO_API(gboolean, mono_unity_method_has_attribute, (MonoMethod* method, MonoClass* attr_klass))
+DO_API(MonoObject*, mono_unity_class_get_attribute, (MonoClass* klass, MonoClass* attr_klass))
```
Comment (joncham) [inline unity/unity-sources/Runtime/Mono/MonoFunctions.h:316]: Is that plan to add these to mono/il2cpp? Or they will use old APIs in Unity?

File: unity/unity-embed-host/CoreCLRHost.cs [hunk 2]
```diff
@@ -281,24 +282,45 @@ public static int array_length([NativeCallbackType("MonoArray*")] IntPtr array)
...
+        Attribute[] attrs = mb.GetCustomAttributes(attr_class.TypeFromHandleIntPtr()).ToArray();
+        return attrs.Length == 0 ? IntPtr.Zero : attrs[0].ToNativeRepresentation();
     }
 
+    [return: NativeCallbackType("MonoObject*")]
+    public static IntPtr unity_field_get_attribute(
+        [NativeCallbackType("MonoClass*")] IntPtr klass,
+        [NativeCallbackType("MonoClassField*")] IntPtr field,
+        [NativeCallbackType("MonoClass*")] IntPtr attr_class)
+        => FieldInfo.GetFieldFromHandle(field.FieldHandleFromHandleIntPtr(), RuntimeTypeHandle.FromIntPtr(klass))
```
Comment (joncham) [inline unity/unity-embed-host/CoreCLRHost.cs]: Why is this one different than the others?

Comment (UnityAlex) [reply unity/unity-embed-host/CoreCLRHost.cs]: Gah! I thought I had converted them all and missed one. I'll fix.

File: unity/unity-sources/Runtime/Mono/MonoFunctions.h (hunk 1 above)
Comment (UnityAlex) [reply unity/unity-sources/Runtime/Mono/MonoFunctions.h:316]: IL2CPP for the most part doesn't seem to support attribute lookup currently so I could see leveraging a C# solution there. Mono I was planning on leaving on the old code path.

**Review by mrvoorhe — APPROVED** [review]

**Pull request merged**